File: wordLadderII.java
import java.util.*;
public class Solution {
    public ArrayList<ArrayList<String>> findLadders(String start, String end, HashSet<String> dict) {
        ArrayList<ArrayList<String>> result = new ArrayList<ArrayList<String>>();
        LinkedList<String> q = new LinkedList<String>();
        q.add(start);
        HashSet<String> visited = new HashSet<String>();
        visited.add(start);
        HashMap<String, ArrayList<String>> m = new HashMap<String, ArrayList<String>>();
        HashMap<String, Integer> level = new HashMap<String, Integer>();
        level.put(start, 1);
        int maxLevel = -1;
        while (!q.isEmpty()) {
            String tmp = q.remove();
            if (maxLevel != -1 && level.get(tmp) > maxLevel) {
                break;
            }
            for (int i=0; i < tmp.length(); i++) {
                for (char c='a'; c <= 'z'; c++) {
                    if (c == tmp.charAt(i)) {
                        continue;
                    }
                    char[] chars = tmp.toCharArray();
                    chars[i] = c;
                    String temp = new String(chars);
                    if (temp.equals(end)) {
                        maxLevel = level.get(tmp);
                        if (!m.containsKey(end)) {
                            m.put(end, new ArrayList<String>());
                        }
                        m.get(end).add(tmp);
                    } else if (dict.contains(temp)) {
                        if (!visited.contains(temp)) {
                            q.add(temp);
                            visited.add(temp);
                            m.put(temp,  new ArrayList<String>());
                            m.get(temp).add(tmp);
                            level.put(temp,  level.get(tmp) + 1);
                        } else if (level.get(tmp)+1 == level.get(temp)) {
                            m.get(temp).add(tmp);
                        }
                    }
                }
            }
        }
        ArrayList<String> path = new ArrayList<String>();
        buildPaths(start, end, m, path, result);
        return result;
    }
    void buildPaths(String start, String end, HashMap<String, ArrayList<String>> m, 
            ArrayList<String> path, ArrayList<ArrayList<String>> result) {
        if (end.equals(start)) {
            ArrayList<String> clone = (ArrayList<String>)path.clone();
            clone.add(0, start);
            result.add(clone);
            return;
        }
        path.add(0, end);
        for (int i=0; m.containsKey(end) && i < m.get(end).size(); ++i) {
            buildPaths(start, m.get(end).get(i), m, path, result);
        }
        path.remove(0);
    }
}

File: Algorithms/wordLadderII.py
class Solution:
    # @param start, a string
    # @param end, a string
    # @param dict, a set of string
    # @return a list of lists of string
    def findLadders(self, start, end, dict):
        self.result, self.path, m = [], [], {}
        q, level, maxLevel = [start], {start:1}, -1
        visited = set()
        visited.add(start)
        while q:
            tmp = q.pop(0)
            if maxLevel != -1 and level[tmp] > maxLevel:
                break;
            for i in range(len(tmp)):
                for c in string.lowercase:
                    if c == tmp[i]: continue
                    temp = tmp[:i] + c + tmp[i+1:]
                    if temp == end:
                        maxLevel = level[tmp]
                        if not end in m:
                            m[end] = []
                        m[end].append(tmp)
                    elif temp in dict:
                        if not temp in visited:
                            q.append(temp)
                            visited.add(temp)
                            m[temp] = [tmp]
                            level[temp] = level[tmp] + 1
                        elif level[tmp]+1 == level[temp]:
                            m[temp].append(tmp)
        self.buildPaths(start, end, m)
        return self.result

    def buildPaths(self, start, end, m):
        if start == end:
            self.result.append([start] + self.path[:])
            return
        self.path.insert(0, end)
        if end in m:
            for i in range(len(m[end])):
                self.buildPaths(start, m[end][i], m)
        self.path.pop(0)
